geo/src/algorithm/line_locate_point.rs

```diff
@@ -41,23 +48,34 @@ impl<T> LineLocatePoint<T, Point<T>> for Line<T>
 where
     T: CoordinateType + Float + Zero + One,
 {
-    type Output = T;
+    type Output = Option<T>;
     type Rhs = Point<T>;
 
     fn line_locate_point(&self, p: &Self::Rhs) -> Self::Output {
+        // let $s$ be the starting point of the line, and $v$ its
+        // direction vector. We want to find $l$ such that 
+        // $(p - (s + lv)) \cdot v = 0$, i.e. the vector from
+        // $l$ along the line to $p$ is perpendicular to $v$.a
+
+        // vector $p - s$
         let sp = [p.x() - self.start.x, p.y() - self.start.y];
+
+        // direction vector of line, $v$
         let v = [self.end.x - self.start.x, self.end.y - self.start.y];
+
+        // $v \cdot v$
         let v_sq = v[0] * v[0] + v[1] * v[1];
         if v_sq == T::zero() {
             // The line has zero length, return zero
-            return T::zero();
+            return Some(T::zero());
         } else {
+            // $v \cdot (p - s)$
             let v_dot_sp = v[0] * sp[0] + v[1] * sp[1];
             let l = v_dot_sp / v_sq;
-            if v_dot_sp.is_nan() | v_sq.is_nan() {
-                return T::nan();
+            if l.is_finite() {
```

Comment:
I think the case I had in my head was where you have an infinitely long line - consider
```
let line = line![(x: -inf, y:0), (x:0, y:0)];
let point = (x: -2, y: 2)
```
In this case both `v_dot_sp` and `v_sq` will be `inf`, and I'm not quite sure how to interpret the fraction along an infinite length line anyway.

But I think your case _is_ sensible, so maybe I only need to guard against lines with infinities in them.

Reply:
I doubt my assumption of how things should work are rigorous. I was just being intuitive. How do others (e.g. postgis) handle this situation?

Reply:
It's not exactly clear from postgis, but it appears they check only whether the line is zero length, and otherwise return the floating point result of `v_dot_sp / v_sq` (maxed/minned to 1/0). This leads to some interesting cases.

```
line: (-1,1) -> (1, 1)
point: (100, inf)
fraction: NaN 
// The dot product of (2, 0) (direction vector) and (101, inf) (point - starting point) is 200 + 0 * inf = 200 + NaN = NaN

line: (-1,1) -> (1, 1 - 0.0000000001)
point: (100, inf)
fraction: 0.0
// The dot product (2, -0.0000000001) and (101, inf) is 200 + (-inf) = -inf

line: (-1,1) -> (1, 1 + 0.0000000001)
point: (100, inf)
fraction: 1.0
// The dot product (2, 0.0000000001) and (101, inf) is 200 + inf = inf
```
So I'm not sure there's much guidance there unfortunately.

Thinking about your example a bit more, I'm having trouble generalizing it - what should the result be for a line `(-10, 0) -> (10, 0)` and a point `(5, inf)`? One might say we should return `0.75` but `(5, inf)` is equally close to `(-10, 0)` and `(10, 0)` if you use a floating-point pythagorus distance - both distances are `inf`. And the dot product of the direction vector and `point - starting point` is NaN.

I'm leaning in the direction that these cases don't really have much meaning, so I think at this stage my preference is either for the following two:

 1. Return `Option<T>` which is `Some` if both point and line have only finite coordinates and `None` otherwise, signifying that locating a point has peculiar semantics (the `Some` variant should only ever contain a finite `T`) (i.e. leaving as is, with some cleaning up).
 2. Return `T` which rolls with the punches of floating points (appears to be the postgis approach)

Either way, a consumer has to think about what to do if their points/lines have non-finite coordinates, but maybe the first option is clearer?

Reply:
After thinking about it for a while... I have no strong opinion. 🤷 

What you have proposed here seems reasonable.